webrtc/README.md

```diff
@@ -0,0 +1,130 @@
+# WebRTC
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active |                 |
+
+Authors: [@mxinden]
+
+Interest Group: [@marten-seemann]
+
+## Motivation
+
+1. **No need for valid TLS certificates.** Enable browsers to connect to public
+   server nodes without those server nodes providing a TLS certificate within
+   the browsers trustchain. Note that we can not do this today with our
+   Websocket transport.
+
+2. **Hole punching in the browser**: Enable two browsers or a browser and a
+   non-public server node to connect.
+
+## Requirements
+
+- Loading a remote nodes certificate into ones browser trust-store is not an
+  option, i.e. doesn't scale.
+
+- No dependency on central STUN and/or TURN servers.
+
+## Protocol
+
+Scenario: Browser _A_ wants to connect to server node _B_.
+
+### Setup - Both _A_ and _B_
+
+1. [Generate a
+   certificate](https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-generatecertificate).
+
+2. [Get the certificate's
+   fingerprint](https://www.w3.org/TR/webrtc/#dom-rtccertificate-getfingerprints).
+
+3. Sign this fingerprint using the libp2p peer key to generate a "peer cert"
+
+### Connection Establishment
+
+1. Browser _A_ discovers server node _B_'s multiaddr, containing _B_'s IP, UDP
+  port, TLS certificate fingerprint and libp2p peer ID (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint/p2p/<peer-id>`),
```

Comment:
(unrelated to my first remark)

What would the listening equivalent of this multiaddress be?
For example, for rust-libp2p, what would you pass to the `Transport::listen_on()` function? Are you expected to pass the certificate fingerprint as well?

Reply:
presently the certificate is passed as an argument. The webrtc-rs API requires you to pass a certificate plus a private key if you want to have a persistent certificate.

```rust
    // --> constructor
    /// Create a new WebRTC transport.
    ///
    /// Creates a UDP socket bound to `listen_addr`.
    pub async fn new<A: ToSocketAddrs>(
        certificate: RTCCertificate,
        listen_addr: A,
    ) -> Result<Self, TransportError<Error>> {
    
    // --> build a config
                config: RTCConfiguration {
                certificates: vec![certificate],
                ..Default::default()
            },
```

Reply:
> Say for example that we add another udp based transport protocol, which one would this address target?

Why the address needs to target anything at all? Sorry if I'm asking stupid questions.

When I'm creating a webrtc transport, I pass the certificate/privatekey pair and IP/PORT as listening address. Then the transport will listen for any webrtc related data on the UDP socket. If someone attempts to bind to the same port, it will fail (because we don't set SO_REUSEPORT). Why IP/PORT has to represent something?

Or this is spec related question? Then I guess the listening URL would be `/ip6/2001:db8::/udp/1234/webrtc/<tls-certificate-fingerprint`. Meaning the peer will listen for any UDP webrtc data and will use the certificate with `<tls-certificate-fingerprint>` fingerprint when establishing a connection.

Reply:
Thanks for explaining! Yep, this can be resolved now.